src/payload.rs

```diff
@@ -0,0 +1,516 @@
+            LogicalClock {
+                id: external_id,
+                epoch,
+                ticks,
+            },
+        )
+        .is_err()
+        {
+            Err(ClocksOverflowedError)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn as_slice(&self) -> &[LogicalClock] {
+        self.0.as_slice()
+    }
+}
+
+pub type DidClocksOverflow = bool;
+pub type NumEntriesRead = usize;
+pub type NumEntriesWritten = usize;
+
+/// Process as much of the given log iterator as possible into a report, outputting the report into output.
+/// Returns: The number of items written to the report, the number of items read from the log, and whether or not
+/// the clocks list overflowed
+///
+/// Note: Assumes that output has enough space to fit all frontier clocks.
+pub fn write_report_payload<I, C, O>(
+    mut log_iter: I,
+    frontier_clocks: &mut C,
+    output: &mut O,
+) -> (NumEntriesWritten, NumEntriesRead, DidClocksOverflow)
+where
+    I: Iterator<Item = PossiblyMissed<LogEntry>>,
+    C: FrontierClocks,
+    O: PayloadOutput,
+{
+    write_clocks(frontier_clocks, output);
+    let mut num_read = 0;
+    let mut num_written = 0;
+    let mut clocks_overflowed = false;
+    while let Some(read_entry) = log_iter.next() {
+        let cur_entry = if let PossiblyMissed::Entry(e) = read_entry {
+            e
+        } else {
+            // Entry was missed, need to count missed and log internal event
+            if output.free_capacity() < 2 {
+                // Could not fit log items missed event
+                return (num_read, num_written, clocks_overflowed);
+            }
+            let mut num_missed = 1;
+            let mut possible_entry_after_missed = None;
+            for possible_missed_entry in &mut log_iter {
+                if let PossiblyMissed::Entry(non_missed_entry) = possible_missed_entry {
+                    possible_entry_after_missed = Some(non_missed_entry);
+                    break;
+                } else {
+                    num_missed += 1
+                }
+            }
+            num_read += num_missed;
+            push_log_items_missed_event(output, num_missed as u32);
+            num_written += 2;
+
+            if let Some(e) = possible_entry_after_missed {
+                e
+            } else {
+                // No entries after missed entries, can simply return
+                // Note: This should only happen in the debug collector, as the RaceBuffer writer
+                // will always have some entries to read after missed entries
+                return (num_written, num_read, clocks_overflowed);
+            }
+        };
+
+        if cur_entry.has_clock_bit_set() {
+            if output.free_capacity() < 2 {
+                // Could not fit clock
+                return (num_read, num_written, clocks_overflowed);
+            }
+            // RaceBuffer guarantees that if prefix is iterated over/read, then suffix will also be available and present
```

Comment:
If this is the case, then we should consider whether the `log_iter` provided to this function should have a different `Item` type that reflects this promise -- e.g. some sort of `enum` that yields either a single non-prefix value or a prefix/suffix pair.  

This may be too much type-attention to ask at this point, but may be a useful thought experiment for a possible future improvement.

Reply:
Yeah, that sounds reasonable to me.